sbws/lib/destination.py

```diff
@@ -245,23 +219,10 @@ def next(self):
         '''
         Returns the next destination that should be used in a measurement
         '''
-        with self._usability_lock:
-            while True:
-                if self._should_perform_usability_test():
-                    self._perform_usability_test()
-                    log.debug('%s/%s of our configured destinations are '
-                              'usable at this time', len(self._usable_dests),
-                              len(self._all_dests))
-                if len(self._usable_dests) > 0:
-                    break
-                time_till_next_check = self._usability_test_interval + 0.0001
-                log.warning(
-                    'Of our %d configured destinations, none are usable at '
-                    'this time. Sleeping %f seconds on this blocking call '
-                    'to DestinationList.next() until we can check for a '
-                    'usable destination again.', len(self._all_dests),
-                    time_till_next_check)
-                time.sleep(time_till_next_check)
-
-        self._rng.shuffle(self._usable_dests)
-        return self._usable_dests[0]
+        # Do not perform usability tests since a destination is already proven
+        # usable or not in every measurement, and it should depend on a X
+        # number of failures.
+        # This removes the need for an extra lock for every measurement.
+        # Do not change the order of the destinations, just return a
+        # destination.
+        return self._rng.choice(self.functional_destinations)
```

Comment:
It looks like sbws will slowly reduce its list of destinations on every network error, until there are none left.
Can a destination ever recover after it is excluded?

There are two cases that we need to handle:
* If a network error is only in one destination, and it lasts for a short time, sbws should use the destination once the network starts working again.
* If a network error affects all destinations, and it lasts for a long time, sbws should error.

Can we retry non-functional destinations with a low probability?

Can we do a few slow retries before failing?
Or can we wait for a few minutes before failing?

Reply:
Created #29589